### PR TITLE
Fix np.where logic in table.join

### PR DESCRIPTION
**Pull request by bsipocz**
This regression was unearthed while doing #9287, and thus be opened as a separate PR to be able to backport the fix.

### CODE REVIEW

File: astropy/table/operations.py [hunk 1]
```diff
@@ -810,11 +815,11 @@ def _join(left, right, keys=None, join_type='inner',
             else:
                 # np.where does not work for mixin columns (e.g. Quantity) so
                 # use a slower workaround.
-                left_mask = ~right_mask
-                if np.any(left_mask):
-                    out[out_name][left_mask] = left[left_name].take(left_out)
+                non_right_mask = ~right_mask
                 if np.any(right_mask):
-                    out[out_name][right_mask] = right[right_name].take(right_out)
+                    out[out_name] = left[left_name].take(left_out)
```
Comment (astrofrog) [inline astropy/table/operations.py]: Why this change? And should this not be an in-place change at least? (with ``[:]``)

Comment (bsipocz) [reply astropy/table/operations.py]: indeed, it should use `[:]`. As for the rest of the changes, I did the fix to follow the very same logic as the `np.where` above (the issue with that it doesn't understand Quantities for np <1.17 and astropy <4.0):
```
                out[out_name][:] = np.where(right_mask,
                                            left[left_name].take(left_out),
                                            right[right_name].take(right_out))
```

**Comment by bsipocz** [comment]
(a rebase has solved the weird appveyor issue, so if we don't see it popping up elsewhere where are good)
